crates/sparrow-compiler/src/types/inference.rs

```diff
@@ -359,7 +381,20 @@ fn instantiate_type(fenl_type: &FenlType, solutions: &HashMap<TypeVariable, Fenl
             let s = Arc::new(Field::new("entries", DataType::Struct(fields), false));
             FenlType::Concrete(DataType::Map(s, false))
         }
-        FenlType::Collection(Collection::List, _) => todo!("unsupported"),
+        FenlType::Collection(Collection::List, type_vars) => {
+            debug_assert!(type_vars.len() == 1);
+            let concrete_type = solutions
+                .get(&type_vars[0])
+                .cloned()
+                .unwrap_or(FenlType::Concrete(DataType::Null));
+            let field = match concrete_type {
+                // TODO: Should the fields be nullable?
```

Comment:
probably. alternatively, it needs to be something we track / propagate.

basically:

1. Non-nullable fields allow certain optimizations. Often, doesn't matter since the kernel looks at the `null_count` anyway, and can use the optimization on an array that is nullable but has no null values or on a non-nullable array.
2. Non-nullable fields cause a panic if a null value is added. I could easily see eventually creating a list from an input that is null... so we should support it.

Reply:
Hm..I'm pretty sure we propagate this now actually. I tested this by changing the nullability and the item name to something other than what is expected in the dataset, and still see the projected schema in the `Scan` set to the dataset's schema:
The `bool_list`'s item was changed to `test` and the nullability to false, which you can see is not here. (This is the correct schema as given by pqrs).
```
2023-07-27T18:40:45.067696Z DEBUG Operation{index=0 operation_label="scan"}: sparrow_runtime::execute::operation: Full operation is ScanOperation { projected_schema: Schema { fields: [Field { name: "time", data_type: Timestamp(Nanosecond, None), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "subsort", data_type: UInt64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "key", data_type: UInt64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "i64_list", data_type: List(Field { name: "item", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "bool_list", data_type: List(Field { name: "item", data_type: Boolean, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "string_list", data_type: List(Field { name: "item", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "index", data_type: Int32, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }], metadata: {} }, key_hash_index: KeyHashIndex { key_hash_to_index: "0 entries" }, .. }
```

Schema:
```
  OPTIONAL group bool_list (LIST) {
    REPEATED group list {
      OPTIONAL BOOLEAN item;
    }
  }

```

Reply:
So this is just a placeholder type that we create during inference, and we verify they match. However..I'm not sure if downstream of this that type is expected rather than the projected schema's type? I haven't seen any issues yet, but perhaps if we have kernels specialized for null and non-null maps/lists we'll see problems if they don't match?